---
title: "05_Ceolin.Rmd"
author: "Rayna M Harris"
date: November 28, 2017
output:
  md_document:
  variant: markdown_github
---

## Reproduction of and comparison to the Ceolin et al. 2017 study.

From [Ceolin L, Bouquier N, Vitre-Boubaker J, Rialle S et al. Cell Type-Specific mRNA Dysregulation in Hippocampal CA1 Pyramidal Neurons of the Fragile X Syndrome Mouse Model. Front Mol Neurosci 2017;10:340. PMID: 29104533](https://www.frontiersin.org/articles/10.3389/fnmol.2017.00340/full)

This data was made available here [open source data](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE94559).

The parts of their paper that I reproduced are visuzalized in Figure 2 and 3 of Ceolin et al 2017. Ceolin's fluoresence staining of the CA1 provided the inspriation for the color palette. I reproduced the heatmap and part of the GO anlsysis but not that of FMRP binding. Then, I compared this "reproduced analysis of the Ceolin data" to my primary results to identify genes that robustly change expression in CA1 following *Fmr1* gene knock out.

![](../figures/fig3-02.png)

I reproduced the data from the Ceolin et al. 2017 which used fluorescence labeling to selectively sequence pyramidal neurons in the CA1 subfield of the hippocampus from WT and FMR1-KO mice (Fig 2A). My reproduction of their data produced a very similar pattern of gene expression and list of differentially expressed genes with roughly equal up and downregulation of expression. Then, I asked how many of their differentially expressed genes are differentially expressed in my study.  I found found that downregulation of expression of Cacna1g, Efcab6, Serpina3n, and Sstr3 was consistent in both the Coelin data and in my data (Fig 2B). Next, I check to see if the genes that I calculated to be significantly different als identified by Coelin and described as significantly different. I determined that 39 of top 45 most significant (p < 0.01) genes in my analysis make up over half of the most significant (p < 0.05) genes of from the Ceolin study (Fig 2C). Of my list of "replicated" 39 differentially expressed genes, two genes (Serpina3a and Efcab6) were also identified in my analysis of differential expression (Fig 2D). My GO analysis highlighted different but also overlapping patterns. The Ceolin study highlights the molecular function enriched pathways in FMR1-KO mice, but my analysis provided stronger evidence for a deletion of calcium receptor-related functions (Fig 2E). This suggests a role for dysregulation of calcium signaling in the hippocampus of Fragile X Syndrome patients and is consistent with my research findings.
 
 

![](../figures/fig3-01.png)
Reproducing the Ceolin study for direct comparison of results.
A) Graphical representation of the samples for the Ceolin et al. 2017 study examining CA1 expression in WT and FMR1-KO mice. B) Reproduction: This volcano plot shows that my analysis of the Ceolin et al count data identified 88 genes that are up-regulated in FMR1-KO mice and the 146 genes that are up-regulated in WT mice a p < 0.05. Comparison: The gene expression and significance values from the Ceolin data are color-coded by the levels of significance from my results described in Fig 2.8. Four genes that are upregulated in WT in my study were also upregulated in my reproduction of the Ceolin data. C) Analysis showing that 39 of top 45 most significant (p < 0.01) genes in my reproduction of the analysis, make up over half of the top most significant (p < 0.05) genes of from the Ceolin study. D) Hierarchical clustering shows the names and expression patterns of those same significant genes. D) GO analysis showing a very similar pattern of depletion of calcium channel activity as was shown in Fig. 2.8). In contrast, Ceolin detected enrichment of ribosomal processes in response to FMR1-KO in CA1 pyramidal neurons. Legend) Teal: Enriched in FMR1-KO, pink: enriched in WT, grey: genes with insignificant expression, black: genes whos expression was not calculated in my original analysis




Here is the analysis, 


```{r setup, message=FALSE,comment=FALSE, warning=FALSE}
#source("http://www.bioconductor.org/biocLite.R")
#biocLite("DESeq2")
library(DESeq2)
library(magrittr)
library(tidyverse)
library(plyr)
library(reshape2)
library(VennDiagram)
library(genefilter)
library(pheatmap)
library(edgeR)
library(colorRamps) # for a matlab like color scheme
library(viridis)
library(genefilter)  ## for PCA fuction
library(ggrepel) ## for labeling volcano plot
library(cowplot)

source("functions_RNAseq.R")

# set output file for figures 
knitr::opts_chunk$set(fig.path = '../figures/05_Ceolin/')
```


```{r ImportData, message=FALSE,comment=FALSE, warning=FALSE}
# contains a file with the gene name and transcript id
geneids <- read.table("../data/geneids.tsv", header=T)

#read count data 
count <- read.table("../data/GSE94559_Raw_Counts_RNA-Seq_Ceolin.txt", header = T)

# strip gene length from data
count$Gene <- sapply(strsplit(as.character(count$Gene),'\\|'), "[", 1)
count$gene <- sapply(strsplit(as.character(count$Gene),'\\|'), "[", 1)

## join with geneids so we can look at gene level stuff
countbygene <- full_join(geneids, count)
countbygene <- countbygene %>% 
  filter(gene != "-")
countbygene <- countbygene[-c(1,3)] ## keep gene name and counts for samples)


## lengthen the dataframe, then wide with gene level sums, then make gene the row name, then round the value to nearest integer
countbygene <- melt(countbygene, id=c("gene")) 
countbygene  <- dcast(countbygene, gene ~ variable, value.var= "value", fun.aggregate=sum)
row.names(countbygene) <- countbygene$gene
countbygene[1] <- NULL
countbygene <- round(countbygene)

# getting ready for DESeq2
countData <- countbygene 


# meta data
sample=c("KO1", "KO2", "KO3", "KO4","KO5", "KO6", 
           "WT1", "WT2","WT3", "WT4", "WT5", "WT6") 
genotype=c("FMR1_KO", "FMR1_KO", "FMR1_KO", "FMR1_KO", "FMR1_KO", "FMR1_KO", 
           "WT", "WT","WT", "WT", "WT", "WT") 

colData <- data.frame(sample,genotype)
rownames(colData) <- colData$sample


# keep only data with >2 counts
countData[countData < 2] <- 0

# replace nas with 0
countData[is.na(countData)] <- 0
```

The first thing I notice is that they have waay more reads per sample and thus gene counts per sample than I do. They have a mean gene counts per sample around 400 million counts per gene. My data had 5 million counts per gene. 

```{r edgeR, message=FALSE,comment=FALSE, warning=FALSE }
counts <- countData
dim( counts )
colSums( counts ) / 1e06  # in millions of reads
table( rowSums( counts ) )[ 1:30 ] # Number of genes with low counts

rowsum <- as.data.frame(colSums( counts ) / 1e06 )
names(rowsum)[1] <- "millioncounts"
rowsum$sample <- row.names(rowsum)

ggplot(rowsum, aes(x=millioncounts)) + 
  geom_histogram(bins = 20, colour = "black", fill = "darkgrey") +
  theme_classic() +
  scale_x_continuous(name = "Millions of Gene Counts per Sample") +
  scale_y_continuous(name = "Number of Samples")
```

Then I conducted differential gene experssion with DESeq2 modeling the effect genotype.

```{r DifferentialGeneExpressionAnalysis}
dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = ~ genotype )
dds <- dds[ rowSums(counts(dds)) > 2, ] ## filter genes with 0 counts
dds <- DESeq(dds) # Differential expression analysis
dds

## for variance stablized gene expression and log transformed data
rld <- rlog(dds, blind=FALSE)

res <- results(dds, contrast =c("genotype", "FMR1_KO", "WT"), independentFiltering = T, alpha = 0.1)
summary(res)
```

There are 12 samples (6 WT and 6 FMR1-KO) and 37,167 genes were included in the analysis, but most were discarded during the normalizaiton and analysis iwth DESeq such that the expression of only 19,361 genes were analysed.

**Differentially expressed genes**

- out of 19361 with nonzero total read count
- adjusted p-value < 0.1
- LFC > 0 (up)     : 88, 0.45% 
- LFC < 0 (down)   : 146, 0.75%

```{r viewcoldata, message=FALSE,comment=FALSE, warning=FALSE}
colData$genotype <- as.factor(colData$genotype)
colData %>% select(genotype)  %>%  summary()

dim(countData)

dim(rld)
```

Then I did my favorite principle component analysis. The  clustering of points shows clear separation of samples by PC1 and PC2 together. PC2 is signfificant.

```{r PCA, message=FALSE,comment=FALSE, warning=FALSE}

# create the dataframe using my function pcadataframe
pcadata <- pcadataframe(rld, intgroup=c("genotype"), returnData=TRUE)
percentVar <- round(100 * attr(pcadata, "percentVar"))

## PC1 vs PC2

PCA12 <- ggplot(pcadata, aes(PC1, PC2, shape = genotype, color = genotype)) + 
  geom_point(size = 3, alpha = 1) +
    xlab(paste0("PC1: ", percentVar[1],"% variance")) +
    ylab(paste0("PC2: ", percentVar[2],"% variance")) +
    scale_color_manual(values =c("#41b6c4", "#e7298a")) +
    theme_cowplot(font_size = 8, line_size = 0.25)  +
    #theme(legend.position="none") +
    scale_shape_manual(values=c(16, 16)) 
PCA12

aov1 <- aov(PC1 ~ genotype, data=pcadata)
summary(aov1) 


aov2 <- aov(PC2 ~ genotype, data=pcadata)
summary(aov2) 

aov3 <- aov(PC3 ~ genotype, data=pcadata)
summary(aov3) 

```

The heatmap shows a similar pattern as the volcano plot and PCA analysis and allows us to visualize patterns of expression with gene names.

```{r HeatmapPadj, message=FALSE,comment=FALSE, warning=FALSE}
contrast1 <- resvals(contrastvector = c('genotype', 'FMR1_KO', 'WT'), mypval = 0.1)

## Any padj <0.1
DEGes <- assay(rld)
DEGes <- cbind(DEGes, contrast1)
DEGes <- as.data.frame(DEGes) # convert matrix to dataframe
DEGes$rownames <- rownames(DEGes)  # add the rownames to the dataframe

DEGes <- DEGes %>% filter(padjgenotypeFMR1_KOWT < 0.01)

rownames(DEGes) <- DEGes$rownames
drop.cols <-colnames(DEGes[,grep("padj|pval|rownames", colnames(DEGes))])
DEGes <- DEGes %>% select(-one_of(drop.cols))
DEGes <- as.matrix(DEGes)
DEGes <- DEGes - rowMeans(DEGes)

# setting color options
ann_colors <- list(genotype =  c('FMR1_KO' = (values=c("#41b6c4")), 
            'WT' = (values=c("#e7298a"))))

df <- as.data.frame(colData(dds)[,c( "genotype")])
rownames(df) <- names(countData)
colnames(df) <- "genotype"

paletteLength <- 40
myBreaks <- c(seq(min(DEGes), 0, length.out=ceiling(paletteLength/2) + 1), 
              seq(max(DEGes)/paletteLength, max(DEGes), length.out=floor(paletteLength/2)))


pheatmap(DEGes, show_colnames=T, show_rownames = T,
         annotation_col=df, 
         annotation_colors = ann_colors,
         treeheight_row = 0, treeheight_col = 10,
         border_color = "grey60" ,
         color = viridis(40), breaks=myBreaks,
         clustering_distance_cols="correlation" ,
         clustering_method="average"
         )

# for adobe
pheatmap(DEGes, show_colnames=F, show_rownames = T,
         annotation_col=df, annotation_colors = ann_colors,
         treeheight_row = 0, treeheight_col = 10,
         fontsize = 4, width=2, height=3.4, cellwidth = 5, 
         border_color = "grey60" ,
         color = viridis(40), breaks=myBreaks,
         clustering_distance_cols="correlation" ,
         clustering_method="average",
         filename = "../figures/05_Ceolin/HeatmapPadj-1.pdf"
         )

```

## Create list of p-values for all genes 

```{r GOsetup, message=FALSE,comment=FALSE, warning=FALSE}
#create a new DF with the gene counts
GOpvals <- assay(rld)
GOpvals <- cbind(GOpvals, contrast1)
GOpvals <- as.data.frame(GOpvals)
GOpvals <- GOpvals[ , grepl( "padj|pval" , names( GOpvals ) ) ]

GOpvals$gene<-rownames(GOpvals)

GOpvals <- GOpvals %>%
  select(gene, padjgenotypeFMR1_KOWT)
GOpvals$logP <- log(GOpvals$padjgenotypeFMR1_KOWT)
GOpvals <- GOpvals %>%
  select(gene, logP)

write.csv(GOpvals, "./06_GO_MWU/05_Ceolin_GOpvals.csv", row.names = F)
```




## Volcanos

```{r volcanos, message=FALSE,comment=FALSE, warning=FALSE}
res <- results(dds, contrast =c("genotype", "FMR1_KO", "WT"), independentFiltering = T, alpha = 0.1)
summary(res)
resOrdered <- res[order(res$padj),]
head(resOrdered, 10)
data <- data.frame(gene = row.names(res), pvalue = -log10(res$padj), lfc = res$log2FoldChange)
data <- na.omit(data)

data$wrap <- "Reproduction"
data <- data %>%
  mutate(color = ifelse(data$lfc > 0 & data$pvalue > 1.3, 
                        yes = "FRM1_KO", 
                        no = ifelse(data$lfc < 0 & data$pvalue > 1.3, 
                                    yes = "WT", 
                                    no = "none")))
top_labelled <- top_n(data, n = 5, wt = pvalue)

topGene <- rownames(res)[which.min(res$padj)]
plotCounts(dds, gene = topGene, intgroup=c("genotype"))

# Color corresponds to fold change directionality

volcano <- ggplot(data, aes(x = lfc, y = pvalue)) + 
  geom_point(aes(color = factor(color)), size = 1, alpha = 0.5, na.rm = T) + # add gene points
  theme_cowplot(font_size = 8, line_size = 0.25) +
  geom_hline(yintercept = 1.3,  size = 0.25, linetype = 2) + 
  scale_color_manual(values = c("FRM1_KO" = "#41b6c4",
                                "WT" = "#e7298a", 
                                "none" = "grey")) + 
  #scale_y_continuous(limits=c(0, 8)) +
  scale_x_continuous(name="Log fold change")+
  ylab(paste0("log10 p-value")) +       
  theme(panel.grid.minor=element_blank(),
        legend.position = "none", # remove legend 
        panel.grid.major=element_blank()) +
  facet_wrap(~wrap)
volcano


pdf(file="../figures/05_Ceolin/volcano.pdf", width=1.25, height=1.85)
plot(volcano)
dev.off()

```

## Venn Diagram of both study's DEGS

```{r venn}
contrast1 <- resvals(contrastvector = c("genotype", "FMR1_KO", "WT"), mypval = 0.01)

#create a new DF with the gene counts
rldpvals <- assay(rld)
rldpvals <- cbind(rldpvals, contrast1)
rldpvals <- as.data.frame(rldpvals)
rldpvals <- rldpvals[ , grepl( "padj|pval" , names( rldpvals ) ) ]
names(rldpvals)

# venn with padj values
venn1 <- row.names(rldpvals[rldpvals[2] <0.01 & !is.na(rldpvals[2]),])
venn2 <- read.csv("../data/GSE94559_Ceolin_DEGS.csv", header = F)
venn2 <- as.matrix(venn2)


candidates <- list("Harris" = venn1, "Ceolin" = venn2)

prettyvenn <- venn.diagram(scaled=T,
  x = candidates, filename=NULL, 
  col = "black",
  fill = c( "white", "white"),
  alpha = 0.5,
  cex = 1, fontfamily = "sans", #fontface = "bold",
  cat.default.pos = "text",
  #cat.dist = c(0.1, 0.1, 0.1), cat.pos = 1,
  cat.cex = 1, 
  cat.fontfamily = "sans")
#dev.off()
grid.draw(prettyvenn)

venn12 <- intersect(venn1,venn2)
write(venn12, "./06_GO_MWU/CeolinHarrisOverlap.csv")
```

# Over reproducible differentially expressed genes 
I took their list of differentially expressed genes at p < 0.05 and my list at 0.01 and identified the overlap. Then I made a heatmap.

```{r heatmapoverlap}
DEGes <- assay(rld)
DEGes <- cbind(DEGes, contrast1)
DEGes <- as.data.frame(DEGes) # convert matrix to dataframe
DEGes$rownames <- rownames(DEGes)  # add the rownames to the dataframe
head(DEGes)

Coelin <- read.csv("../data/GSE94559_Ceolin_DEGS.csv", header = F)
colnames(Coelin)<- c("rownames")
DEGes <- inner_join(DEGes, Coelin)
DEGes <- DEGes %>% filter(rownames != "Col1a1")

DEGes <- DEGes %>% filter(padjgenotypeFMR1_KOWT < 0.01)

rownames(DEGes) <- DEGes$rownames
drop.cols <-colnames(DEGes[,grep("padj|pval|rownames", colnames(DEGes))])
DEGes <- DEGes %>% select(-one_of(drop.cols))
DEGes <- as.matrix(DEGes)
DEGes <- DEGes - rowMeans(DEGes)
head(DEGes)

# setting color options
ann_colors <- list(genotype =  c('FMR1_KO' = (values=c("#41b6c4")), 
            'WT' = (values=c("#e7298a"))))

df <- as.data.frame(colData(dds)[,c( "genotype")])
rownames(df) <- names(countData)
colnames(df) <- "genotype"

paletteLength <- 40
myBreaks <- c(seq(min(DEGes), 0, length.out=ceiling(paletteLength/2) + 1), 
              seq(max(DEGes)/paletteLength, max(DEGes), length.out=floor(paletteLength/2)))


pheatmap(DEGes, show_colnames=T, show_rownames = T,
         annotation_col=df, 
         annotation_colors = ann_colors,
         treeheight_row = 0, treeheight_col = 10,
         border_color = "grey60" ,
         color = viridis(40), breaks=myBreaks,
         clustering_distance_cols="correlation" ,
         clustering_method="average"
         )

pheatmap(DEGes, show_colnames=F, show_rownames = T,
         annotation_col=df, annotation_colors = ann_colors,
         treeheight_row = 0, treeheight_col = 0,
         fontsize = 5, width=2, height=3.4, cellwidth = 5, 
         border_color = "grey60" ,
         color = viridis(40), breaks=myBreaks,
         clustering_distance_cols="correlation" ,
         clustering_method="average",
         filename = "../figures/05_Ceolin/HeatmapOverlap.pdf"
         )


```

Suzy-like volcano plot

```{r suzyvolcano}
pointcolor <- read.csv("../results/FMR1_CA1_rnaseq.csv")

res <- results(dds, contrast =c("genotype", "FMR1_KO", "WT"), independentFiltering = T, alpha = 0.1)
summary(res)
resOrdered <- res[order(res$padj),]
head(resOrdered, 10)
data <- data.frame(gene = row.names(res), pvalue = -log10(res$padj), lfc = res$log2FoldChange)
data <- na.omit(data)

tempdata <- left_join(data, pointcolor, by = "gene")
tempdata$color <- as.character(tempdata$color)
tempdata$color[is.na(tempdata$color)] <- "absent"
tempdata$color <- as.factor(tempdata$color)
levels(tempdata$color) <- list(FMR1KO="FMR1", WT="WT", NS = "none",absent="absent")

tempdata$wrap <- "Comparison"


# Color corresponds to fold change directionality

suzyvolcano <- ggplot(tempdata, aes(x = lfc.x, y = pvalue.x)) + 
  geom_point(aes(color = color, shape = color, size = color), alpha = 0.5, na.rm = T) + # add gene points
  theme_cowplot(font_size = 8, line_size = 0.25) +
  geom_hline(yintercept = 1.3,  size = 0.25, linetype = 2) + 
  scale_color_manual(values = c("FMR1KO" = "#41b6c4",
                                "WT" = "#e7298a", 
                                "NS" = "grey",
                                "absent" = "black")) + 
  #scale_y_continuous(limits=c(0, 8)) +
  scale_x_continuous(name="Log fold change")+
  scale_shape_manual(values=c(16, 16, 16, 16)) +
  scale_size_manual(values=c(2,2, 0.5, 0.5)) +
 scale_y_continuous(name=NULL,
                    labels = NULL)+      
  theme(panel.grid.minor=element_blank(),
        legend.position = "none", # remove legend 
        panel.grid.major=element_blank()) +
  facet_wrap(~wrap)
suzyvolcano

pdf(file="../figures/05_Ceolin/suzyvolcano.pdf", width=1, height=1.85)
plot(suzyvolcano)
dev.off()

legendvolcano <- ggplot(tempdata, aes(x = lfc.x, y = pvalue.x)) + 
  geom_point(aes(color = color, shape = color, size = color), alpha = 0.5, na.rm = T) + # add gene points
  theme_cowplot(font_size = 9, line_size = 0.25) +
  geom_hline(yintercept = 1.3,  size = 0.25, linetype = 2) + 
  scale_color_manual(values = c("FMR1KO" = "#41b6c4",
                                "WT" = "#e7298a", 
                                "NS" = "grey",
                                "absent" = "black")) + 
  #scale_y_continuous(limits=c(0, 8)) +
  scale_x_continuous(name="Log fold change")+
  scale_shape_manual(values=c(16, 16, 16, 16)) +
  scale_size_manual(values=c(8,8, 1, 1)) +
 scale_y_continuous(name=NULL,
                    labels = NULL)+      
  theme(panel.grid.minor=element_blank(),
        #legend.position = "none", # remove legend 
        panel.grid.major=element_blank())
legendvolcano

pdf(file="../figures/05_Ceolin/legendvolcano.pdf", width=1, height=1.75)
plot(legendvolcano)
dev.off()

# list top deges
mytopgenes <- tempdata %>%
  filter(color %in% c("WT", "FMR1"), pvalue.x > 1.3)
head(mytopgenes)
mytopgenes$gene

```